ENIGMAsystem/SHELL/Audio_Systems/SDL/SDLadvanced.cpp

```diff
@@ -0,0 +1,142 @@
+#include "SoundResource.h"
+#include "SoundChannel.h"
+#include "SDLsystem.h"
+#include "../General/ASadvanced.h"
+#include "Audio_Systems/audio_mandatory.h"
+#include "Widget_Systems/widgets_mandatory.h"
+#include "../General/ASutil.h"
+#include <SDL_mixer.h>
+#include <string>
+#include <stdio.h>
+
+const int AUDIO_CHANNEL_OFFSET = 400000;
+const int AUDIO_CHANNEL_COUNT = 128;
+vector<SoundChannel *> sound_channels;
+
+namespace enigma_user {
+
+bool audio_exists(int index) { return sounds.exists(index); }
+
+int audio_add(string fname) {
+ // Open sound
+  size_t flen = 0;
+  char* fdata = enigma::read_all_bytes(fname, flen);
+  if (!fdata) {
+    DEBUG_MESSAGE(std::string("The Sound file - ") + std::string(fname) + std::string(" failed to open"), MESSAGE_TYPE::M_ERROR);
+    return -1;
+  }
+  // Decode sound
+  int rid = sounds.size();
+  bool fail = enigma::sound_add_from_buffer(rid, fdata, flen);
+  delete[] fdata;
+  
+  if (fail) {
+    DEBUG_MESSAGE(std::string("The Sound file - ") + std::string(fname) + std::string(" failed to be decoded"), MESSAGE_TYPE::M_ERROR);
+    return -1;
+  }
+  return rid;
+}
+
+int audio_play_sound(int index, double priority, bool loop) {
+  const Sound& snd = sounds.get(index);
+  if (sound_channels.empty() || sound_channels.size() <= AUDIO_CHANNEL_COUNT - 1) {
+    SoundChannel* sc = new SoundChannel();
+    sc->mchunk = snd.mc;
+    sc->priority = priority;
+    sc->soundIndex = index;
+    sound_channels.push_back(sc);
+    if (Mix_PlayChannel(sound_channels.size() -1,snd.mc, loop) == -1) {
+      DEBUG_MESSAGE( std::string("Mix_PlayChannel : ")+std::string(Mix_GetError()), MESSAGE_TYPE::M_ERROR);

```

Comment:
Don't use `Mix_GetError` as we talked about, it's unreliable.
https://www.libsdl.org/projects/SDL_mixer/docs/SDL_mixer_9.html#SEC9
>"Note: this function does not always set the error string, so do not depend on Mix_GetError being meaningful all the time."
```suggestion
      DEBUG_MESSAGE( std::string("Mix_PlayChannel failed: ")+std::to_string(index), MESSAGE_TYPE::M_ERROR);
```